### PR TITLE
Add -Z maximal-hir-to-mir-coverage flag

### PR DESCRIPTION
**Pull request by willcrichton**
This PR adds a new unstable flag `-Z maximal-hir-to-mir-coverage` that changes the behavior of `maybe_lint_level_root_bounded`, pursuant to [a discussion on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Mapping.20MIR.20to.20HIR). When enabled, this function will not search upwards for a lint root, but rather immediately return the provided HIR node ID. This change increases the granularity of the mapping between MIR locations and HIR nodes inside the `SourceScopeLocalData` data structures. This increase in granularity is useful for rustc consumers like [Flowistry](https://github.com/willcrichton/flowistry) that rely on getting source-mapping information about the MIR CFG that is as precise as possible.

A test `maximal_mir_to_hir_coverage.rs` has been added to verify that this flag does not break anything.

r? @cjgillot 

cc @gavinleroy

### CODE REVIEW

File: compiler/rustc_middle/src/lint.rs [hunk 1]
```diff
@@ -173,6 +173,16 @@ impl TyCtxt<'_> {
...
+        // Some consumers of rustc need to map MIR locations back to HIR nodes. Currently the
+        // the only part of rustc that tracks MIR -> HIR is the `SourceScopeLocalData::lint_root`
+        // field that tracks lint levels for MIR locations.  Normally the number of source scopes
+        // is limited to the set of nodes with lint  annotations. The -Zmaximal-hir-to-mir-coverage
+        // flag changes this behavior to maximize the number of source scopes, increasing the
+        // granularity of the MIR->HIR mapping.
+        if self.sess.opts.unstable_opts.maximal_hir_to_mir_coverage {
+            return id;
+        }
+
```
Comment (cjgillot) [inline compiler/rustc_middle/src/lint.rs:185]: The name of this function does not hint any relationship to MIR. Could you prefer using this flag in `rustc_mir_build` 
There are only 2 places where it's used, with the same pattern. Refactoring them to a single `maybe_create_source_scope` that checks this option may be the better option.

File: compiler/rustc_mir_build/src/build/scope.rs [hunk 2]
```diff
@@ -579,13 +580,13 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
             );
             let current_root = tcx.maybe_lint_level_root_bounded(current_hir_id, self.hir_id);
```
Comment (cjgillot) [inline compiler/rustc_mir_build/src/build/scope.rs:581]: Those 2 calls should go into `maybe_new_source_scope`.

Comment (willcrichton) [reply compiler/rustc_mir_build/src/build/scope.rs]: Ah yes I was just fixing this when you left his comment :-)

**Review by cjgillot — COMMENTED** [review]
@bors r+

**Comment by willcrichton** [comment]
https://github.com/willcrichton/flowistry/commit/5eb8f457e953c1b009e0b197adf1769b7dded590

Net -140 lines of code! Thanks @cjgillot!!!

**Pull request merged**